exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,1848 @@
+	if a == asset.Empty {
+		a = ok.GuessAssetTypeFromInstrumentID(response.Argument.InstrumentID)
+	}
+	pair, err = ok.GetPairFromInstrumentID(response.Argument.InstrumentID)
+	if err != nil {
+		pair.Delimiter = currency.DashDelimiter
+	}
+	for i := range response.Data {
+		if response.Action == OkxOrderBookSnapshot ||
+			response.Argument.Channel == okxChannelOrderBooks5 ||
+			response.Argument.Channel == okxChannelBBOTBT {
+			err = ok.WsProcessSnapshotOrderBook(response.Data[i], pair, a)
+			if err != nil {
+				_, err2 := ok.OrderBooksSubscription("subscribe", response.Argument.Channel, a, pair)
+				if err2 != nil {
+					ok.Websocket.DataHandler <- err2
+				}
+				return err
+			}
+			if ok.Verbose {
+				log.Debugf(log.ExchangeSys,
+					"%s passed checksum for pair %v",
+					ok.Name, pair,
+				)
+			}
+		} else {
+			if len(response.Data[i].Asks) == 0 && len(response.Data[i].Bids) == 0 {
+				return nil
+			}
+			err := ok.WsProcessUpdateOrderbook(response.Argument.Channel, response.Data[i], pair, a)
+			if err != nil {
+				_, err2 := ok.OrderBooksSubscription("subscribe", response.Argument.Channel, a, pair)
+				if err2 != nil {
+					ok.Websocket.DataHandler <- err2
+				}
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// WsProcessSnapshotOrderBook processes snapshot order books
+func (ok *Okx) WsProcessSnapshotOrderBook(data WsOrderBookData, pair currency.Pair, a asset.Item) error {
+	var err error
+	if data.Checksum != 0 {
+		var signedChecksum int32
+		signedChecksum, err = ok.CalculateOrderbookChecksum(data)
+		if err != nil {
+			return fmt.Errorf("%s channel: Orderbook unable to calculate orderbook checksum: %s", ok.Name, err)
+		}
+		if signedChecksum != data.Checksum {
+			return fmt.Errorf("%s channel: Orderbook for %v checksum invalid",
+				ok.Name,
+				pair)
+		}
+	}
+
+	asks, err := ok.AppendWsOrderbookItems(data.Asks)
+	if err != nil {
+		return err
+	}
+	bids, err := ok.AppendWsOrderbookItems(data.Bids)
+	if err != nil {
+		return err
+	}
+	newOrderBook := orderbook.Base{
+		Asset:           a,
+		Asks:            asks,
+		Bids:            bids,
+		LastUpdated:     data.Timestamp,
+		Pair:            pair,
+		Exchange:        ok.Name,
+		VerifyOrderbook: ok.CanVerifyOrderbook,
+	}
+	return ok.Websocket.Orderbook.LoadSnapshot(&newOrderBook)
+}
+
+// WsProcessUpdateOrderbook updates an existing orderbook using websocket data
+// After merging WS data, it will sort, validate and finally update the existing
+// orderbook
+func (ok *Okx) WsProcessUpdateOrderbook(channel string, data WsOrderBookData, pair currency.Pair, a asset.Item) error {
+	update := &orderbook.Update{
+		Asset: a,
+		Pair:  pair,
+	}
+	switch channel {
+	case okxChannelOrderBooks,
+		okxChannelOrderBooksTBT:
+		update.MaxDepth = 400
+	case okxChannelOrderBooks5:
+		update.MaxDepth = 5
+	case okxChannelBBOTBT:
+		update.MaxDepth = 1
+		update.MaxDepth = 400
+	case okxChannelOrderBooks50TBT:
+		update.MaxDepth = 50
+	}
```

Comment:
Remove as this has a chance to dynamically cleave the entire stored tranche levels if there is multiple book types streamed.